### ISSUE TITLE
Crashed when calling asyncio.run(coro)

### ISSUE THREAD

**Issue by dimaqq** [ipython/ipython#12000]
#### Versions
* `Python-3.8.0`
* `ipython==7.10.1`

#### Environment
* `PYTHONASYNCIODEBUG=1`

#### Steps to reproduce

```
(x-wNEHA9fb-py3.8) … ~/x> ipython
Python 3.8.0 (v3.8.0:fa919fdf25, Oct 14 2019, 10:23:27)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.10.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: %config Application.verbose_crash=True

In [2]: import asyncio

In [3]: async def foo():
   ...:     print(1)
   ...:     await asyncio.sleep(1)
   ...:     print(2)
   ...:

In [4]: async def bar():
   ...:     x = foo()
   ...:     print(3)
   ...:     await asyncio.sleep(2)
   ...:     print(4)
   ...:

In [5]: asyncio.run(bar())
3
4
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/events.py:81: RuntimeWarning: coroutine 'foo' was never awaited
  self._context.run(self._callback, *self._args)
RuntimeWarning: Enable tracemalloc to get the object allocation traceback

---------------------------------------------------------------------------
```

#### Crash report
```
***************************************************************************

IPython post-mortem report

{'commit_hash': 'e508b6e4d',
 'commit_source': 'installation',
 'default_encoding': 'utf-8',
 'ipython_path': '/Users/u/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/IPython',
 'ipython_version': '7.10.1',
 'os_name': 'posix',
 'platform': 'macOS-10.15.1-x86_64-i386-64bit',
 'sys_executable': '/Users/u/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/bin/python',
 'sys_platform': 'darwin',
 'sys_version': '3.8.0 (v3.8.0:fa919fdf25, Oct 14 2019, 10:23:27) \n'
                '[Clang 6.0 (clang-600.0.57)]'}

***************************************************************************



***************************************************************************

Crash traceback:

---------------------------------------------------------------------------
---------------------------------------------------------------------------
RuntimeErrorPython 3.8.0: /Users/u/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/bin/python
                                                   Mon Dec  2 15:34:29 2019
A problem occurred executing Python code.  Here is the sequence of function
calls leading up to the error, with the most recent (innermost) call last.
~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/bin/ipython in <module>
      1 #!/Users/u/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/bin/python
      2 # -*- coding: utf-8 -*-
      3 import re
      4 import sys
      5 from IPython import start_ipython
      6 if __name__ == '__main__':
      7     sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])
----> 8     sys.exit(start_ipython())
        global sys.exit = <built-in function exit>
        global start_ipython = <function start_ipython at 0x7feaa06483a0>

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/IPython/__init__.py in start_ipython(argv=None, **kwargs={})
    111     This is a public API method, and will survive implementation changes.
    112
    113     Parameters
    114     ----------
    115
    116     argv : list or None, optional
    117         If unspecified or None, IPython will parse command-line options from sys.argv.
    118         To prevent any command-line parsing, pass an empty list: `argv=[]`.
    119     user_ns : dict, optional
    120         specify this dictionary to initialize the IPython user namespace with particular values.
    121     kwargs : various, optional
    122         Any other kwargs will be passed to the Application constructor,
    123         such as `config`.
    124     """
    125     from IPython.terminal.ipapp import launch_new_instance
--> 126     return launch_new_instance(argv=argv, **kwargs)
        launch_new_instance = <bound method Application.launch_instance of <class 'IPython.terminal.ipapp.TerminalIPythonApp'>>
        argv = None
        kwargs = {}
    127
    128 def start_kernel(argv=None, **kwargs):
    129     """Launch a normal IPython kernel instance (as opposed to embedded)
    130
    131     `IPython.embed_kernel()` puts a shell in a particular calling scope,
    132     such as a function or method for debugging purposes,
    133     which is often not desirable.
    134
    135     `start_kernel()` does full, regular IPython initialization,
    136     including loading startup files, configuration, etc.
    137     much of which is skipped by `embed()`.
    138
    139     Parameters
    140     ----------
    141

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/traitlets/config/application.py in launch_instance(cls=<class 'IPython.terminal.ipapp.TerminalIPythonApp'>, argv=None, **kwargs={})
    649             lines.append(cls.class_config_section())
    650         return '\n'.join(lines)
    651
    652     def exit(self, exit_status=0):
    653         self.log.debug("Exiting application: %s" % self.name)
    654         sys.exit(exit_status)
    655
    656     @classmethod
    657     def launch_instance(cls, argv=None, **kwargs):
    658         """Launch a global instance of this Application
    659
    660         If a global instance already exists, this reinitializes and starts it
    661         """
    662         app = cls.instance(**kwargs)
    663         app.initialize(argv)
--> 664         app.start()
        app.start = <bound method TerminalIPythonApp.start of <IPython.terminal.ipapp.TerminalIPythonApp object at 0x7feac0267130>>
    665
    666 #-----------------------------------------------------------------------------
    667 # utility functions, for convenience
    668 #-----------------------------------------------------------------------------
    669
    670 def boolean_flag(name, configurable, set_help='', unset_help=''):
    671     """Helper for building basic --trait, --no-trait flags.
    672
    673     Parameters
    674     ----------
    675
    676     name : str
    677         The name of the flag.
    678     configurable : str
    679         The 'Class.trait' string of the trait to be set/unset with the flag

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/IPython/terminal/ipapp.py in start(self=<IPython.terminal.ipapp.TerminalIPythonApp object>)
    341         if self.log_level <= logging.INFO: print()
    342
    343     def _pylab_changed(self, name, old, new):
    344         """Replace --pylab='inline' with --pylab='auto'"""
    345         if new == 'inline':
    346             warnings.warn("'inline' not available as pylab backend, "
    347                       "using 'auto' instead.")
    348             self.pylab = 'auto'
    349
    350     def start(self):
    351         if self.subapp is not None:
    352             return self.subapp.start()
    353         # perform any prexec steps:
    354         if self.interact:
    355             self.log.debug("Starting IPython's mainloop...")
--> 356             self.shell.mainloop()
        self.shell.mainloop = <bound method TerminalInteractiveShell.mainloop of <IPython.terminal.interactiveshell.TerminalInteractiveShell object at 0x7feaa062a910>>
    357         else:
    358             self.log.debug("IPython not interactive...")
    359             if not self.shell.last_execution_succeeded:
    360                 sys.exit(1)
    361
    362 def load_default_config(ipython_dir=None):
    363     """Load the default config file from the default ipython_dir.
    364
    365     This is useful for embedded shells.
    366     """
    367     if ipython_dir is None:
    368         ipython_dir = get_ipython_dir()
    369
    370     profile_dir = os.path.join(ipython_dir, 'profile_default')
    371     app = TerminalIPythonApp()

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/IPython/terminal/interactiveshell.py in mainloop(self=<IPython.terminal.interactiveshell.TerminalInteractiveShell object>, display_banner=<object object>)
    537                 if (not self.confirm_exit) \
    538                         or self.ask_yes_no('Do you really want to exit ([y]/n)?','y','n'):
    539                     self.ask_exit()
    540
    541             else:
    542                 if code:
    543                     self.run_cell(code, store_history=True)
    544
    545     def mainloop(self, display_banner=DISPLAY_BANNER_DEPRECATED):
    546         # An extra layer of protection in case someone mashing Ctrl-C breaks
    547         # out of our internal code.
    548         if display_banner is not DISPLAY_BANNER_DEPRECATED:
    549             warn('mainloop `display_banner` argument is deprecated since IPython 5.0. Call `show_banner()` if needed.', DeprecationWarning, stacklevel=2)
    550         while True:
    551             try:
--> 552                 self.interact()
        self.interact = <bound method TerminalInteractiveShell.interact of <IPython.terminal.interactiveshell.TerminalInteractiveShell object at 0x7feaa062a910>>
    553                 break
    554             except KeyboardInterrupt as e:
    555                 print("\n%s escaped interact()\n" % type(e).__name__)
    556             finally:
    557                 # An interrupt during the eventloop will mess up the
    558                 # internal state of the prompt_toolkit library.
    559                 # Stopping the eventloop fixes this, see
    560                 # https://github.com/ipython/ipython/pull/9867
    561                 if hasattr(self, '_eventloop'):
    562                     self._eventloop.stop()
    563
    564                 self.restore_term_title()
    565
    566
    567     _inputhook = None

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/IPython/terminal/interactiveshell.py in interact(self=<IPython.terminal.interactiveshell.TerminalInteractiveShell object>, display_banner=<object object>)
    520     def ask_exit(self):
    521         self.keep_running = False
    522
    523     rl_next_input = None
    524
    525     def interact(self, display_banner=DISPLAY_BANNER_DEPRECATED):
    526
    527         if display_banner is not DISPLAY_BANNER_DEPRECATED:
    528             warn('interact `display_banner` argument is deprecated since IPython 5.0. Call `show_banner()` if needed.', DeprecationWarning, stacklevel=2)
    529
    530         self.keep_running = True
    531         while self.keep_running:
    532             print(self.separate_in, end='')
    533
    534             try:
--> 535                 code = self.prompt_for_code()
        code = 'asyncio.run(bar())'
        self.prompt_for_code = <bound method TerminalInteractiveShell.prompt_for_code of <IPython.terminal.interactiveshell.TerminalInteractiveShell object at 0x7feaa062a910>>
    536             except EOFError:
    537                 if (not self.confirm_exit) \
    538                         or self.ask_yes_no('Do you really want to exit ([y]/n)?','y','n'):
    539                     self.ask_exit()
    540
    541             else:
    542                 if code:
    543                     self.run_cell(code, store_history=True)
    544
    545     def mainloop(self, display_banner=DISPLAY_BANNER_DEPRECATED):
    546         # An extra layer of protection in case someone mashing Ctrl-C breaks
    547         # out of our internal code.
    548         if display_banner is not DISPLAY_BANNER_DEPRECATED:
    549             warn('mainloop `display_banner` argument is deprecated since IPython 5.0. Call `show_banner()` if needed.', DeprecationWarning, stacklevel=2)
    550         while True:

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/IPython/terminal/interactiveshell.py in prompt_for_code(self=<IPython.terminal.interactiveshell.TerminalInteractiveShell object>)
    437                         filter=HasFocus(DEFAULT_BUFFER) & ~IsDone() &
    438                             Condition(lambda: self.highlight_matching_brackets))],
    439                 }
    440         if not PTK3:
    441             options['inputhook'] = self.inputhook
    442
    443         return options
    444
    445     def prompt_for_code(self):
    446         if self.rl_next_input:
    447             default = self.rl_next_input
    448             self.rl_next_input = None
    449         else:
    450             default = ''
    451
--> 452         with patch_stdout(raw=True):
        global patch_stdout = <function patch_stdout at 0x7feac0875d30>
        global raw = undefined
    453             # In order to make sure that asyncio code written in the
    454             # interactive shell doesn't interfere with the prompt, we run the
    455             # prompt in a different event loop.
    456             # If we don't do this, people could spawn coroutine with a
    457             # while/true inside which will freeze the prompt.
    458
    459             old_loop = asyncio.get_event_loop()
    460             asyncio.set_event_loop(self.pt_loop)
    461             try:
    462                 text = self.pt_app.prompt(
    463                     default=default,
    464                     **self._extra_prompt_options())
    465             finally:
    466                 # Restore the original event loop.
    467                 asyncio.set_event_loop(old_loop)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/contextlib.py in __enter__(self=<contextlib._GeneratorContextManager object>)
     98                                AbstractContextManager,
     99                                ContextDecorator):
    100     """Helper for @contextmanager decorator."""
    101
    102     def _recreate_cm(self):
    103         # _GCM instances are one-shot context managers, so the
    104         # CM must be recreated each time a decorated function is
    105         # called
    106         return self.__class__(self.func, self.args, self.kwds)
    107
    108     def __enter__(self):
    109         # do not keep args and kwds alive unnecessarily
    110         # they are only needed for recreation, which is not possible anymore
    111         del self.args, self.kwds, self.func
    112         try:
--> 113             return next(self.gen)
        global next = undefined
        self.gen = <generator object patch_stdout at 0x7fea900b5200>
    114         except StopIteration:
    115             raise RuntimeError("generator didn't yield") from None
    116
    117     def __exit__(self, type, value, traceback):
    118         if type is None:
    119             try:
    120                 next(self.gen)
    121             except StopIteration:
    122                 return False
    123             else:
    124                 raise RuntimeError("generator didn't stop")
    125         else:
    126             if value is None:
    127                 # Need to force instantiation so we can reliably
    128                 # tell if we get the same exception back

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/prompt_toolkit/patch_stdout.py in patch_stdout(raw=True)
     32
     33
     34 @contextmanager
     35 def patch_stdout(raw: bool = False) -> Generator[None, None, None]:
     36     """
     37     Replace `sys.stdout` by an :class:`_StdoutProxy` instance.
     38
     39     Writing to this proxy will make sure that the text appears above the
     40     prompt, and that it doesn't destroy the output from the renderer.  If no
     41     application is curring, the behaviour should be identical to writing to
     42     `sys.stdout` directly.
     43
     44     :param raw: (`bool`) When True, vt100 terminal escape sequences are not
     45                 removed/escaped.
     46     """
---> 47     proxy = cast(TextIO, StdoutProxy(raw=raw))
        proxy = undefined
        global cast = <function cast at 0x7feac048c550>
        global TextIO = <class 'typing.TextIO'>
        global StdoutProxy = <class 'prompt_toolkit.patch_stdout.StdoutProxy'>
        raw = True
     48
     49     original_stdout = sys.stdout
     50     original_stderr = sys.stderr
     51
     52     # Enter.
     53     sys.stdout = proxy
     54     sys.stderr = proxy
     55
     56     try:
     57         yield
     58     finally:
     59         # Exit.
     60         proxy.flush()
     61
     62         sys.stdout = original_stdout

~/Library/Caches/pypoetry/virtualenvs/x-wNEHA9fb-py3.8/lib/python3.8/site-packages/prompt_toolkit/patch_stdout.py in __init__(self=<prompt_toolkit.patch_stdout.StdoutProxy object>, raw=True, original_stdout=<_io.TextIOWrapper name='<stdout>' mode='w' encoding='utf-8'>)
     73         self, raw: bool = False, original_stdout: Optional[TextIO] = None
     74     ) -> None:
     75
     76         original_stdout = original_stdout or sys.__stdout__
     77
     78         self.original_stdout = original_stdout
     79
     80         self._lock = threading.RLock()
     81         self._raw = raw
     82         self._buffer: List[str] = []
     83
     84         # errors/encoding attribute for compatibility with sys.__stdout__.
     85         self.errors = original_stdout.errors
     86         self.encoding = original_stdout.encoding
     87
---> 88         self.loop = get_event_loop()
        self.loop = undefined
        global get_event_loop = <built-in function get_event_loop>
     89
     90     def _write_and_flush(self, text: str) -> None:
     91         """
     92         Write the given text to stdout and flush.
     93         If an application is running, use `run_in_terminal`.
     94         """
     95         if not text:
     96             # Don't bother calling `run_in_terminal` when there is nothing to
     97             # display.
     98             return
     99
    100         def write_and_flush() -> None:
    101             self.original_stdout.write(text)
    102             self.original_stdout.flush()
    103

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/events.py in get_event_loop(self=<asyncio.unix_events._UnixDefaultEventLoopPolicy object>)
    624
    625     def __init__(self):
    626         self._local = self._Local()
    627
    628     def get_event_loop(self):
    629         """Get the event loop for the current context.
    630
    631         Returns an instance of EventLoop or raises an exception.
    632         """
    633         if (self._local._loop is None and
    634                 not self._local._set_called and
    635                 isinstance(threading.current_thread(), threading._MainThread)):
    636             self.set_event_loop(self.new_event_loop())
    637
    638         if self._local._loop is None:
--> 639             raise RuntimeError('There is no current event loop in thread %r.'
        global RuntimeError = undefined
        global threading.current_thread.name = undefined
    640                                % threading.current_thread().name)
    641
    642         return self._local._loop
    643
    644     def set_event_loop(self, loop):
    645         """Set the event loop."""
    646         self._local._set_called = True
    647         assert loop is None or isinstance(loop, AbstractEventLoop)
    648         self._local._loop = loop
    649
    650     def new_event_loop(self):
    651         """Create a new event loop.
    652
    653         You must call set_event_loop() to make this the current event
    654         loop.

RuntimeError: There is no current event loop in thread 'MainThread'.

***************************************************************************

History of session input:get_ipython().run_line_magic('config', 'Application.verbose_crash=True')import asyncioasync def foo():
    print(1)
    await asyncio.sleep(1)
    print(2)
    async def bar():
    x = foo()
    print(3)
    await asyncio.sleep(2)
    print(4)
    asyncio.run(bar())
*** Last line of input (may not be in above history):
asyncio.run(bar())
```

**Comment by dimaqq** [comment]
Ugh, it seems any `asyncio.run(...)` causes this.
I guess I didn't read the memo that I should just `await bar()` starting from version x.y.z 🤔

**Comment by Carreau** [comment]
I'm going to guess you are under prompt_toolkit 3.0.2 , is that correct ? 
It might work under prmopt toolkit 2.x.

It's a bit tricky as IPython/prompt toolkit also need the event loop , and ingeneral you can't have 2 event loop running at the same time. 

We should definitively not crash; but it might be hard to do automatically the right thing.

**Comment by Carreau** [comment]
The simplest version of this is
```
import asyncio
asyncio.run(asyncio.sleep(0))
```

The reason seem to be that `asyncio.run()` close the event loop and prompt toolkit it using it. 

Not sure why, if there is an eventloop running, it should refuse to run... 

@jonathanslenders may have some ideas.

**Comment by Carreau** [comment]
minimal pure python example:

```
from asyncio import run, sleep, get_event_loop

print(get_event_loop()) # return a value
run(sleep(0))
print(get_event_loop()) # raise. 
```

Might be a python bug ?

**Comment by Carreau** [comment]
opened https://bugs.python.org/issue38955

**Comment by jonathanslenders** [comment]
It could be a bug in prompt_toolkit, I'm looking into. Not sure yet. 

Use ``asyncio.get_event_loop().run_until_complete(...)`` instead of ``asyncio.run(...)`` for now.

**Comment by jonathanslenders** [comment]
Well, I think we have to fix it in IPython.

It looks like ``get_event_loop()`` will always raise ``RuntimeError`` after having used ``asyncio.run()``. Probably because ``asyncio.run()`` will remove the event loop when done, and that's probably by design.

I *think* it's fine for a REPL to check for this condition and install a new event loop. Maybe the top-level ``await`` in IPython is also broken because of this after a ``asyncio.run()``.

I'll push a fix into this PR: https://github.com/ipython/ipython/pull/12001 because it'll touch the same code.

**Comment by jonathanslenders** [comment]
The second commit in https://github.com/ipython/ipython/pull/12001/commits should fix it.

**Comment by dimaqq** [comment]
Yes, that's correct, `prompt-toolkit==3.0.2`

**Comment by dimaqq** [comment]
> opened https://bugs.python.org/issue38955

Some context in case it's useful:

So, Python disallows nested event loops, see e.g. https://mail.python.org/archives/list/async-sig@python.org/thread/5D6AI56Y3CHLVJHSJRBX5USDDPA5IRNI/

I'm not entirely clear on the semantics, does `get_event_loop()` (truthy) necessarily mean that current code runs under event loop (event loop is running), or may it also mean that current code is synchronous and an event loop has been set (but is not running)?

If it's the earlier, then IMO `.run(...)` should just fail.
If the latter is the case here, then I dunno 🤔

**Comment by dimaqq** [comment]
Based on the error, it's the latter in ipython, `.get_event_loop()` and `.run()` are called from synchronous code, then the current event loop is not running.

The old reference is marked "not running, not closed" after `.run()`... because `.run()` whips up a new (ephemeral) event loop, which is used and then closed. So messy 🤪

**Comment by jonathanslenders** [comment]
Yes, IPython uses prompt_toolkit for the terminal UI, which uses asyncio for its event loop.
We don't want the user to interfere with that event loop, because this could freeze the UI, people could schedule blocking code in the loop using ``asyncio.ensure_future(some_async_func())``.

So, this is why we swap the current event loop with another one, just for prompting, and we restore the event loop afterwards. That should be fine, because it's indeed done from a synchronous context, without any event loop running. The problem we had here is that apparently ``get_event_loop()`` can apparently raise ``RuntimeError``. This is fixed in. https://github.com/ipython/ipython/pull/12001/commits

For the asyncio inputhook, there I made the mistake of using nested event loops, which was unreliable. With the same PR, that's fixed. That means if you do ``%gui asyncio`` in IPython, the UI will run on the same event loop, which means that any asyncio code keeps running in the background, when the prompt shows.

**Comment by Carreau** [comment]
closed by #12001
